.github/workflows/run_periodic_tests.yml

```
@@ -24,22 +24,96 @@ concurrency:
+        os: [ubuntu-latest, macos-12, macos-14, windows-latest]
+        python-version: ["3.9", "3.10", "3.11", "3.12"]
+        # Exclude Python 3.12 from unit tests since we run it in the coverage jobs
+        exclude:
+          - os: ubuntu-latest
+            python-version: "3.12"
+    steps:
+      - name: Check out PyBaMM repository
+        uses: actions/checkout@v4
+
+      - name: Install Linux system dependencies
+        if: matrix.os == 'ubuntu-latest'
+        run: |
+          sudo apt-get update
+          sudo apt-get install gfortran gcc graphviz pandoc libopenblas-dev texlive-latex-extra dvipng
+
+      - name: Install macOS system dependencies
+        if: matrix.os == 'macos-12' || matrix.os == 'macos-14'
+        env:
+          HOMEBREW_NO_INSTALL_CLEANUP: 1
+          HOMEBREW_NO_AUTO_UPDATE: 1
+          HOMEBREW_NO_COLOR: 1
+          # Speed up CI
+          NONINTERACTIVE: 1
+        # sometimes gfortran cannot be found, so reinstall gcc just to be sure
+        run: |
+          brew analytics off
+          brew install graphviz
+          brew reinstall gcc
+
+      - name: Install Windows system dependencies
+        if: matrix.os == 'windows-latest'
+        run: choco install graphviz --version=8.0.5
+
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Install nox
+        run: python -m pip install nox
+
+      - name: Install SuiteSparse and SUNDIALS on GNU/Linux and macOS
+        timeout-minutes: 10
+        if: matrix.os != 'windows-latest'
+        run: python -m nox -s pybamm-requires
+
+      - name: Run unit tests for ${{ matrix.os }} with Python ${{ matrix.python-version }}
+        run: python -m nox -s unit
+
+  check_coverage:
     runs-on: ubuntu-latest
+    name: Coverage tests (ubuntu-latest / Python 3.12)
+
     steps:
-      - uses: actions/checkout@v4
-      - name: Setup python
+      - name: Check out PyBaMM repository
+        uses: actions/checkout@v4
+
+      - name: Install Linux system dependencies
+        run: |
+          sudo apt-get update
+          sudo apt-get install gfortran gcc graphviz pandoc libopenblas-dev texlive-latex-extra dvipng
+
+      - name: Set up Python 3.12
+        id: setup-python
         uses: actions/setup-python@v5
         with:
           python-version: 3.12
 
-      - name: Check style
-        run: |
-          python -m pip install pre-commit
-          pre-commit run -a
+      - name: Install nox
+        run: python -m pip install nox
 
-  build:
-    needs: style
+      - name: Install SuiteSparse and SUNDIALS on GNU/Linux
+        timeout-minutes: 10
+        run: python -m nox -s pybamm-requires
+
+      - name: Run unit tests for Ubuntu with Python 3.12 and generate coverage report
+        run: python -m nox -s coverage
+
+      - name: Upload coverage report
+        uses: codecov/codecov-action@v4.3.1
```

```suggestion
      - name: Upload coverage report
        uses: codecov/codecov-action@v4.3.1
        if: github.repository == 'pybamm-team/PyBaMM'
```

Based on the [warning](https://github.com/cringeyburger/PyBaMM/actions/runs/9162868476/job/25190764132#step:8:37) generated in the coverage job. This doesn't fail the workflow on the error but converts it to a warning – maybe we can future-proof that on the occasion of a later version where they might make this an error after all